README.md
# Scratchpad for [Clojure for the Brave and True]

[Clojure for the Brave and True]: http://www.braveclojure.com/

## License

This software can be used freely, see [The Unlicense](./LICENSE.txt).

## Warranty

This work is provided "as is" and without any express or
implied warranties, including, without limitation, the implied
warranties of merchantibility and fitness for a particular
purpose.


fwpd/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2015-12-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2015-12-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/fwpd/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/fwpd/compare/0.1.0...0.1.1


clojure-noob/src/clojure_noob/hobbits.clj
(ns clojure-noob.hobbits
  (:gen-class))

(def asym-hobbit-body-parts
  [{:name "head" :size 3}
   {:name "left-eye" :size 1}
   {:name "left-ear" :size 1}
   {:name "mouth" :size 1}
   {:name "nose" :size 1}
   {:name "neck" :size 2}
   {:name "left-shoulder" :size 3}
   {:name "left-upper-arm" :size 3}
   {:name "chest" :size 10}
   {:name "back" :size 10}
   {:name "left-forearm" :size 3}
   {:name "abdomen" :size 6}
   {:name "left-kidney" :size 1}
   {:name "left-hand" :size 2}
   {:name "left-knee" :size 2}
   {:name "left-thigh" :size 4}
   {:name "left-lower-leg" :size 3}
   {:name "left-achilles" :size 1}
   {:name "left-foot" :size 2}])

(defn matching-part
  [part]
  {:name (clojure.string/replace (:name part) #"^left-" "right-")
   :size (:size part)})

(defn symmetrize-body-parts
  "Expects a seq of maps that have :name and :size"
  [asym-body-parts]
  (loop [remaining-asym-parts asym-body-parts
         final-body-parts []]
    (if (empty? remaining-asym-parts)
      final-body-parts
      (let [[part & remaining] remaining-asym-parts]
        (recur remaining
               (into final-body-parts (set [part (matching-part part)])))))))

(symmetrize-body-parts asym-hobbit-body-parts)

(defn better-symmetrize-body-parts
  "Expects a seq of maps that have :name and :size"
  [asym-body-parts]
  (reduce #(into %1 (set [%2 (matching-part %2)]))
          [] asym-body-parts))

(better-symmetrize-body-parts asym-hobbit-body-parts)

(defn hit
  "Hit a hobbit"
  [asym-body-parts]
  (let [sym-parts (better-symmetrize-body-parts asym-body-parts)
        body-part-size-sum (reduce + (map :size sym-parts))
        target (rand body-part-size-sum)]
    (loop [[part & remaining] sym-parts
           accumulated-size (:size part)]
      (if (> accumulated-size target)
        part
        (recur remaining (+ accumulated-size (:size (first remaining))))))))

(hit asym-hobbit-body-parts)


clojure-noob/src/clojure_noob/exercises.clj
(ns clojure-noob.exercises
  (:gen-class))

; 3.1
(str "Clap" " Hands")
(vector 1 2 3)
(list :a :b :c)
(hash-map :a 1 :b 2)
(hash-set 1 2 3 3)

; 3.2
(defn add-hundred
  "Add 100 to input"
  [x]
  (+ x 100))
(add-hundred 1)

; 3.3
(defn inc-maker
  "Make an incrementor"
  [n]
  #(+ % n))
(def inc9 (inc-maker 9))
(inc9 10)

; 3.4
(defn mapset
  "Map over a set"
  [f x]
  (set (map f x)))
(mapset inc [1 1 2 2])

; 3.5
(def asym-alien-parts
  [{:name "head" :size 3}
   {:name "left-eye" :size 1}
   {:name "left-ear" :size 1}
   {:name "mouth" :size 1}
   {:name "nose" :size 1}
   {:name "neck" :size 2}
   {:name "left-shoulder" :size 3}
   {:name "left-upper-arm" :size 3}
   {:name "chest" :size 10}
   {:name "back" :size 10}
   {:name "left-forearm" :size 3}
   {:name "abdomen" :size 6}
   {:name "left-kidney" :size 1}
   {:name "left-hand" :size 2}
   {:name "left-knee" :size 2}
   {:name "left-thigh" :size 4}
   {:name "left-lower-leg" :size 3}
   {:name "left-achilles" :size 1}
   {:name "left-foot" :size 2}])

(defn matching-part
  [part loc]
  {:name (clojure.string/replace (:name part) #"^left-" (str loc "-"))
   :size (:size part)})

(defn matching-alien-parts
  [part]
  (map #(matching-part part %)
       ["right" "top" "bottom-left" "bottom-right"]))

(defn symmetrize-body-parts
  "Expects a seq of maps that have :name and :size"
  [asym-body-parts]
  (reduce #(into %1 (set (into [%2] (matching-alien-parts %2))))
          [] asym-body-parts))

(symmetrize-body-parts asym-alien-parts)

; 3.6
; Uses asym-alien-parts and matching-part from 3.5.
(defn matching-n-parts
  [part part-names]
  (map #(matching-part part %) part-names))

(defn symmetrize-n-body-parts
  "Expects a seq of maps that have :name and :size"
  [asym-body-parts part-names]
  (reduce #(into %1 (set (into [%2] (matching-n-parts %2 part-names))))
          [] asym-body-parts))

(symmetrize-n-body-parts asym-alien-parts ["a" "b" "c"])

; Chapter 4 exercises are in the fwpd project.


clojure-noob/src/clojure_noob/core.clj
(ns clojure-noob.core
  (:gen-class))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "I'm a little teapot!"))

; Chapter 3

(+ 1 2 3)

(str "I " "am " "brave.")

(if true
  "Puppies"
  "No Puppies")

(if false
  "Puppies"
  "No Puppies")

(if false
  "No Puppies")

(if true
  (do (println "Success!")
      "Puppies")
  (do (println "Failure!")
      "No Puppies"))

(when true
  (println "Success!")
  "Puppies")

(nil? 1)
(nil? nil?)

(if 0
  "Puppies")

(if nil
  "Puppies"
  "No Puppies")

(= 1 1)
(= nil nil)
(= 1 2)

(or false nil :puppies :kitties)
(or (= 0 1) (= "yes" "no"))
(or nil)

(and :kitties :puppies)
(and :puppies nil false)

(def tools-to-use
  ["vim", "zsh", "tmux", "git"])
tools-to-use

(defn error-message
  [severity]
  (str "Oh noes! "
       (if (= severity :cuddles)
         "Too many puppies!"
         "No puppies!")))

(error-message :cuddles)
(error-message :itsbad)

(hash-map :a 1 :b 2)
(get {:a 0 :b 1} :b)
(get {:a 0 :b 1} :c)
(get {:a 0 :b 1} :c "puppies")
(get-in {:a 0 :b {:c "puppies"}} [:b :c])
({:name "Razor"} :name)

(:a {:a 1 :b 2 :c 3})
(:d {:a 1 :b 2 :c 3} "Nope")

(get [3 2 1] 0)
(get [:a "b" 3] 1)
(vector 3 2 1)
(conj [1 2 3] 4)

'(1 2 3 4)
(nth '(:a :b :c) 0)
(nth '(:a :b :c) 2)
(list 1 "two" {3 4})
(conj '(1 2 3) 4)

#{"kurt vonnegut" 20 :ice 9}
(hash-set 1 1 2 2)
(conj #{:a :b} :b)
(set [3 3 3 4 4])
(contains? #{:a :b} :a)
(contains? #{:a :b} 3)
(contains? #{:a :b nil} nil)
(:a #{:a :b})
(get #{:a :b} :a)
(get #{:a nil} nil)
(get #{:a nil} :c)

(or + -)
((or + -) 1 2 3)
((and (= 1 1) +) 1 2 3)
((first [+ 0]) 1 2 3)

(inc 1.1)
(map inc [0 1 2 3])

(defn too-cute
  "Return a cheer for something cute"
  [thing]
  (str "That " thing "is too cute!"))

(too-cute "puppy")

(defn x-chop
  "Neochop"
  ([name chop]
   (str "Your " chop " chop inflicts pain on " name "."))
  ([name]
   (x-chop name "karate")))

(x-chop "Khan")
(x-chop "Khan" "slap")

(defn squee
  [thing]
  (str "SQUEE! It's a " thing))

(defn squeer
  [& things]
  (map squee things))

(squeer "Puppy" "Kitty")

(defn favorite-things
  [name & things]
  (str "My name is " name " and I like "
       (clojure.string/join ", " things)
       ", etc."))

(favorite-things "Lilly" "puppies" "kitties")

(defn x-first
  [[first-thing]]
  first-thing)

(x-first [1 2 3 4])

(defn x-choose
  [[first-thing second-thing & other-things]]
  (str first-thing ", " second-thing ", "
       (clojure.string/join ", " other-things)))

(x-choose ["this" "that" "the other" "the another"])

(defn x-find
  [{x :horz y :vert}]
  (println (str "(" x ", " y ")")))

(x-find {:horz 3 :vert -7})

(defn x-finder
  [{:keys [x y]}]
  (println (str "(" x ", " y ")")))

(x-finder {:x 4 :y -2})

(defn x-goer
  [{:keys [x y] :as coor}]
  (println (str "(" x ", " y ")"))
  (println coor))

(x-goer {:x 4 :y 2})

(map (fn [name] (str "Hey, " name))
     ["Dick" "Jane"])

(def three-er (fn [x] (* 3 x)))
(three-er 11)

(#(* % 3) 8)

(map #(str "Hey, " %)
     ["Dick" "Jane"])

(#(str %1 " and " %2) "this" "that")
(#(identity %&) 1 2 3 :yo)

(defn inc-maker
  "Make a custom incrementor"
  [inc-by]
  #(+ % inc-by))

(def inc42 (inc-maker 42))
(inc42 8)

(def x 0)
(let [x 1] x)
(let [x (inc x)] x)

(def puppy-list ["Pongo" "Corgi" "Husky"])
(let [[pongo & puppies] puppy-list]
  [pongo puppies])

(into [:b] (set [:a :a]))

(loop [itr 0]
  (println (str "Num " itr))
  (if (> itr 3)
    (println "Done")
    (recur (inc itr))))

; Chapter 4

(seq '(1 2 3))
(seq [1 2 3])
(seq #{1 2 3})
(seq {:a 1 :b 2})
(into {} (seq {:a 1 :b 2}))

(map str ["a" "b" "c"] ["d" "e" "f"])
(map + [1 2 3] [-3 2 1])

(def humans   [2.3 4.5 2.4 6.6 7.7])
(def critters [4.5 6.7 4.3 7.5 3.2])
(defn unify-diet [human critter] {:human human :critter critter})
(map unify-diet humans critters)

(def sum #(reduce + %))
(def avg #(/ (sum %) (count %)))
(defn stats [numbers] (map #(% numbers) [sum count avg]))
(stats [3 4 1])
(stats [3.2 44 53.1 3.11])

(def identities
  [{:alias "Batman" :real "Bruce Wayne"}
   {:alias "Spider-Man" :real "Peter Parker"}
   {:alias "Santa" :real "Your mom"}
   {:alias "Easter Bunny" :real "Your dad"}])

(map :real identities)

(reduce (fn [new-map [key val]]
          (assoc new-map key (inc val)))
        {}
        {:a 20 :b 30})

(reduce (fn [new-map [key val]]
          (if (> val 4)
            (assoc new-map key val)
            new-map))
        {}
        {:a 5 :b 2})

(take 3 [1 2 3 4 5])
(drop 3 [1 2 3 4 5])

(def food-journal
  [{:month 1 :day 1 :human 5.3 :critter 2.3}
   {:month 1 :day 2 :human 5.1 :critter 2.0}
   {:month 2 :day 1 :human 4.9 :critter 2.1}
   {:month 2 :day 2 :human 5.0 :critter 2.5}
   {:month 3 :day 1 :human 4.2 :critter 3.3}
   {:month 3 :day 2 :human 4.0 :critter 3.8}
   {:month 4 :day 1 :human 3.7 :critter 3.9}
   {:month 4 :day 2 :human 3.7 :critter 3.6}])

(take-while #(< (:month %) 3) food-journal)
(drop-while #(< (:month %) 3) food-journal)
(take-while #(< (:month %) 4) (drop-while #(< (:month %) 2) food-journal))
(filter #(< (:human %) 5) food-journal)
(filter #(< (:month %) 3) food-journal)
(some #(> (:critter %) 5) food-journal)
(some #(> (:critter %) 3) food-journal)
(some #(and (> (:critter %) 3) %) food-journal)

(sort [4 3 5 3 2])
(sort-by count ["aaa" "c" "cc"])
(concat [1 2] [3 4])

(concat (take 8 (repeat "na")) ["Batman!"])
(take 3 (repeatedly #(rand-int 10)))

(defn even-numbers
  ([] (even-numbers 0))
  ([n] (cons n (lazy-seq (even-numbers (+ n 2))))))
(take 10 (even-numbers))

(map identity {:a "b"})
(into {} (map identity {:a "b"}))
(into [] (map identity {:a "b"}))
(into [] (map identity [:a "b"]))
(into #{} (map identity [:a :b :a]))
(into {:a 1} [[:b 2]])
(into [:a] '(:b :c))
(into {:a 1} {:b 2})
(conj [0] 1 2 3 4 5)

(max 0 4 5 2)
(apply max [3 5 3 2 45 4 32])
(def add10 (partial + 10))
(add10 20)

(defn lousy-logger
  [log-level message]
  (condp = log-level
    :warn (clojure.string/lower-case message)
    :emergency (clojure.string/upper-case message)))

(def warn (partial lousy-logger :warn))
(warn "Puppies")

; Chapter 5

((comp inc *) 3 5)

(def character {:name "Link" :attrs {:int 10 :str 12 :dex 18}})
(def c-int (comp :int :attrs))
(def c-str (comp :str :attrs))
(def c-dex (comp :dex :attrs))
(c-int character)
(c-str character)
(c-dex character)

(defn spell-slots
  [char]
  (int (inc (/ (c-int char) 2))))

(def spell-slots-comp (comp int inc #(/ % 2) c-int))
(spell-slots-comp character)

(defn sleepy-ident [x] (Thread/sleep 100) x)
(sleepy-ident "Ant-Man")
(sleepy-ident "Ant-Man")

(def memo-sleppy-ident (memoize sleepy-ident))
(memo-sleppy-ident "Ant-Man")
(memo-sleppy-ident "Ant-Man")
